File: src/heap/BinaryHeap.java
package heap;

import java.util.*;
//https://programmers.co.kr/learn/courses/30/lessons/42628
public class BinaryHeap {
    private static class BinaryComparator implements Comparator<Integer>{
        private int direction = 1;
        private Queue<Integer> parentQueue;
        private void changeToASC(){
            if(this.direction == 1) return;
            direction=1;
            refreshParentQueue();
        }
        private void changeToDESC(){
            if(this.direction == -1) return;
            direction = -1;
            refreshParentQueue();
        }
        private void refreshParentQueue(){
            int size = parentQueue.size();
            for(int i = 0; i < size; i++) parentQueue.add(parentQueue.poll());
        }

        void setParentQueue(Queue<Integer> parentQueue) {
            this.parentQueue = parentQueue;
        }

        @Override
        public int compare(Integer o1, Integer o2) {
            return direction*(o1-o2);
        }
    }

    private static class BinaryPriorityQueue extends PriorityQueue<Integer> {
        BinaryPriorityQueue(BinaryComparator comparator) {
            super(comparator);
        }

        private void pollMax(){
            ((BinaryComparator)comparator()).changeToDESC();
            poll();
        }
        private void pollMin(){
            ((BinaryComparator)comparator()).changeToASC();
            poll();
        }
    }

    public int[] solution(String[] operations) {
        int[] answer = {0,0};
        BinaryComparator comparator = new BinaryComparator();
        BinaryPriorityQueue q = new BinaryPriorityQueue(comparator);
        comparator.setParentQueue(q);
        for (String operation : operations) {
            String []cmd = operation.split(" ");
            String cmd1 = cmd[0];
            int cmd2 = Integer.parseInt(cmd[1]);
            switch (cmd1){
                case "D":
                    if(q.isEmpty()) continue;
                    switch (cmd2){
                        case 1:
                            q.pollMax();
                            break;
                        case -1:
                            q.pollMin();
                            break;
                    }
                    break;
                case "I":
                    q.add(cmd2);
                    break;
            }
        }
        if(q.isEmpty()) return answer;
        answer[0] = Collections.max(q);
        answer[1] = Collections.min(q);
        return answer;
    }
    public static void main(String[] args) {
       String[] operations = {"I 19", "I 18", "I 17","I 5","I 5","D 1","D -1"};
        System.out.println(Arrays.toString(new BinaryHeap().solution(operations)));

    }
}

File: src/heap/binary_heap.py
# https://programmers.co.kr/learn/courses/30/lessons/42628?language=python3
# 이중우선순위큐
import heapq
def solution(operations):
    answer = []
    q = []
    for operation in operations:
        oper, operand = operation.split(' ')
        operand = int(operand)
        if oper == 'I':
            heapq.heappush(q, operand)
        else:
            if not q:
                continue
            if operand == -1:
                heapq.heappop(q)
            else:
                q.remove(max(q))
    if q:
        return [max(q), min(q)]
    else:
        return [0,0]
    return answer


if __name__ == '__main__':
    print(solution(["I 16", "D 1"]))
    print(solution(["I 7", "I 5", "I -5", "D -1"]))
    print(solution(["I 16", "I -5643", "D -1", "D 1", "D 1", "I 123", "D -1"]))
    print(solution(["I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"]))
    print(solution(["I 1", "I 2", "I 3", "I 4", "I 5", "I 6", "I 7", "I 8", "I 9", "I 10", "D 1", "D -1", "D 1", "D -1", "I 1", "I 2", "I 3", "I 4", "I 5", "I 6", "I 7", "I 8", "I 9", "I 10", "D 1", "D -1", "D 1", "D -1"]))
